### ISSUE TITLE
try_into_query() returns private error type that cannot be converted to Box<dyn std::error::Error>

### ISSUE THREAD

**Issue by brianmay** [influxdb-rs/influxdb-rust#188]
Disclaimer: I used AI to help write this report. It seems to be correct. Not sure I 100% agree with its solution of implementing std::error::Error would be sufficient to fix though, it should be possible to use a concrete Error type.

## Description:

In influxdb-rs v0.8.0, the try_into_query() method (generated by the #[derive(InfluxDbWriteable)] macro) returns Result<WriteQuery, Self::Error>. However, Self::Error is defined inside a private module generated by the derive macro, making it unusable in external code.
This makes it impossible to properly handle errors without using unwrap():

```
use chrono::{DateTime, Utc};
use influxdb::InfluxDbWriteable;
#[derive(InfluxDbWriteable)]
struct TestReading {
    value: f64,
    time: DateTime<Utc>,
}
fn main() {
    let reading = TestReading { value: 1.0, time: Utc::now() };
    let result: Result<influxdb::WriteQuery, _> = reading.try_into_query("test");
    
    match result {
        Ok(query) => println!("Query: {:?}", query),
        Err(e) => {
            // ERROR: cannot convert to Box<dyn std::error::Error>
            let boxed: Box<dyn std::error::Error> = Box::new(e);
        }
    }
}
```

## Compiler error:

```
error[E0277]: `influxdb::error::TimeTryFromError<...>` doesn't implement `std::fmt::Display`
  --> src/main.rs:22:53
   |
22 |             let boxed: Box<dyn std::error::Error> = Box::new(e);
   |                                                     ^^^^^^^^^^^ unsatisfied trait bound
```

## Root Cause

The derive macro in influxdb_derive-0.6.0/src/writeable.rs generates:
1. A private module named __influxdb_private:
      mod __influxdb_private {
       pub enum Error<T> { ... }
   }
   
2. An impl that sets Self::Error to the private type:
      impl InfluxDbWriteable for MyStruct {
       type Error = __influxdb_private::Error<DateTime<Utc>>;
       // ...
   }
   
## The problem is:

- __influxdb_private::Error<T> contains TimeTryFromError<TimestampTooLargeError, TryFromIntError> as a variant
- TimeTryFromError and TimestampTooLargeError (defined in influxdb/src/error.rs) only implement thiserror::Error, NOT std::error::Error
- Therefore __influxdb_private::Error<DateTime<Utc>> doesn't implement std::error::Error
- And cannot be converted to Box<dyn std::error::Error>

## Expected Behavior 

Users should be able to handle errors from try_into_query() using ? or map_err() in their own error handling patterns.

## Suggested Fix

Either:
1. Make the error type public - Re-export __influxdb_private::Error from the crate so users can name it, OR
2. Implement std::error::Error for the underlying error types - Add std::error::Error impl to TimestampTooLargeError and TimeTryFromError in influxdb/src/error.rs:
      // Current - only thiserror::Error
   #[derive(Clone, Copy, Debug, Error)]
   #[error("The timestamp is too large to fit into an i64.")]
   pub struct TimestampTooLargeError(pub(crate) ());
   
   // Needs: impl std::error::Error for TimestampTooLargeError
   
## Workaround

Users must currently use unwrap() or add #[allow(clippy::unwrap_used)].

## Environment:
- influxdb-rs: v0.8.0
- influxdb_derive: v0.6.0  
- Feature: chrono enabled
- Rust: 1.94

**Comment by msrd0** [comment]
https://github.com/influxdb-rs/influxdb-rust/blob/main/influxdb_derive/src/writeable.rs#L271

Thanks for wasting my time with your hallucinating AI

**Comment by brianmay** [comment]
Yes, you are correct, the public Error type does implement the Error trait.

My report does not concern the public Error type.

Please do read my report, and not discount it just because I mentioned AI. I spent a lot of time checking and rechecking everything it said.

reading.try_into_query() returns the private Error type - __influxdb_private::Error - which is a different type. This private type does not implement the Error trait, your reference doesn't prove anything here.

If you do try running that test code, you do clearly get the mentioned error.

```
use chrono::{DateTime, Utc};
use influxdb::InfluxDbWriteable;
#[derive(InfluxDbWriteable)]
struct TestReading {
    value: f64,
    time: DateTime<Utc>,
}
fn main() {
    let reading = TestReading { value: 1.0, time: Utc::now() };
    let result: Result<influxdb::WriteQuery, _> = reading.try_into_query("test");
    
    match result {
        Ok(query) => println!("Query: {:?}", query),
        Err(e) => {
            // ERROR: cannot convert to Box<dyn std::error::Error>
            let boxed: Box<dyn std::error::Error> = Box::new(e);
        }
    }
}
```

I do get:

```
error[E0277]: `TimeTryFromError<TimestampTooLargeError, ...>` doesn't implement `std::fmt::Display`
  --> src/main.rs:22:53
   |
22 |             let boxed: Box<dyn std::error::Error> = Box::new(e);
   |                                                     ^^^^^^^^^^^ the trait `std::fmt::Display` is not implemented for `TimeTryFromError<TimestampTooLargeError, ...>`
   |
   = help: the trait `std::error::Error` is implemented for `__influxdb_private::Error<T>`
   = note: required for `TimeTryFromError<TimestampTooLargeError, ...>` to implement `std::error::Error`
   = note: 1 redundant requirement hidden
   = note: required for `__influxdb_private::Error<DateTime<Utc>>` to implement `std::error::Error`
   = note: required for the cast from `Box<__influxdb_private::Error<DateTime<Utc>>>` to `Box<dyn std::error::Error>`
   = note: the full name for the type has been written to '/tmp/test_influx/target/debug/deps/test_influx-0549311f236d4293.long-type-2220446071701436319.txt'
   = note: consider using `--verbose` to print the full type name to the console
```

Which clearly shows that the private error type - __influxdb_private::Error does not implement the Error trait.

As I also stated, even if the Error trait is implemented, I don't think that is sufficient, it should be possible to return the specific error type. Which means it can't be private.

**Comment by brianmay** [comment]
Actually I have to wonder why does reading.try_into_query() have to return a private Error type, why can't it just use influxdb::Error which is public?

**Comment by brianmay** [comment]
Possibly all of this is somehow connected to using `DateTime<Utc>` as the time field. Not sure here though, was getting lost trying to understand this.

**Comment by msrd0** [comment]
> reading.try_into_query() returns the private Error type - __influxdb_private::Error - which is a different type. This private type does not implement the Error trait, your reference doesn't prove anything here.

The code linked above is exactly that private error type. And it's not entirely private, since you can name it via `<YourStruct as InfluxdbWriteable>::Error`.

It's error implementation depends on your timestamp type. It looks like you are using `chrono::DateTime`. The error type of it is `TimeTryFromError<TimestampTooLargeError, <u128 as TryFrom<i64>>::Error>`. Both parameters should implement `Error`.

So no idea why an error implementation would be missing.

> > TimeTryFromError and TimestampTooLargeError (defined in influxdb/src/error.rs) only implement thiserror::Error, NOT std::error::Error  
>
> Please do read my report, and not discount it just because I mentioned AI.

Yeah exactly. Maybe teach your AI what thiserror does before it writes a wall of text containing a complete bullshit analysis of the code.

**Comment by brianmay** [comment]
If I missed incorrect AI stuff apologies. Like I said, was trying to understand exactly what the problem was myself.

**Comment by msrd0** [comment]
If you are using any fancy `Box<dyn Error>` wrappers such as `anyhow`, you can work around this by using `.map_err(|err| anyhow!("{err:?}"))` (unlike suggested by your AI that wrote that using `.map_err` was somehow impossible).

**Comment by msrd0** [comment]
Nvm you don't even need anyhow, here's your example with all those "private" error types:

```rust
use chrono::{DateTime, Utc};
use influxdb::{InfluxDbWriteable, Timestamp};

#[derive(InfluxDbWriteable)]
struct TestReading {
	value: f64,
	time: DateTime<Utc>
}

type TestReadingError = <TestReading as InfluxDbWriteable>::Error;
type TimeTryFromError = <Timestamp as TryFrom<DateTime<Utc>>>::Error;

fn main() {
	let reading = TestReading {
		value: 1.0,
		time: Utc::now()
	};
	let result: Result<influxdb::WriteQuery, _> = reading.try_into_query("test");

	match result {
		Ok(query) => println!("Query: {:?}", query),
		Err(e) => {
			// Work around https://github.com/influxdb-rs/influxdb-rust/issues/189
			let boxed: Box<dyn std::error::Error> = match e {
				TestReadingError::TimestampError(TimeTryFromError::TimeError(e)) => Box::new(e),
				TestReadingError::TimestampError(TimeTryFromError::IntError(e)) => Box::new(e),
				TestReadingError::QueryError(e) => Box::new(e)
			};
		}
	}
}
```

**Comment by brianmay** [comment]
Away from computer now, but will try this ASAP.

Some of this could be better documented, reason I look for AI when I can't find documention for the issues I encounter.

**Comment by msrd0** [comment]
Thing is - putting _I get this error when using `?`, no idea why, I don't understand it_ into an issue would've been so much better than pasting that AI nonsense

**Comment by brianmay** [comment]
It started off as "I want to return an Error from this function but the type appears to be private". And I tried to work out why the type was private.

And I wanted to be sure it really was a bug, and not in my code. Which is why I tried diagnosing it. I have had equal number of complaints in the past from people who assume it is some problem with my code without reading it properly :-(

From the code you have above, it appears that I should be able to use TestReadingError - <TestReading as InfluxDbWriteable>::Error - as per the type definition - as the return type, will try it.

OK, I guess the InfluxDbWriteable derive implements the InfluxDbWriteable trait, which provides the Error property. Kind of obvious now that I think about it. :-)

I would suggest that the example code be modified to make use of TestReadingError so you can see how to do Error handling.

And of course #189 is a real problem also. Thanks for making this into easy to understand bug report.

**Comment by brianmay** [comment]
It is a bit of a pain I think that every type has a different Error type. Probably nothing that can't be resolved without generics, does make dealing with this less convenient.

And not sure I can think of a good reason why to do it this way. But that might be reflective of my state of mind right now also :-)

Will look at this again tomorrow.
